cmd/maestro/server/pulse_server.go

```diff
@@ -0,0 +1,113 @@
+// checks all Maestro instances' liveness and trigger statusresync for dead instances.
+type PulseServer struct {
+	instanceID    string
+	pulseInterval int64
+	checkInterval int64
+	instanceDao   dao.InstanceDao
+	lockFactory   db.LockFactory
+	sourceClient  cloudevents.SourceClient
+}
+
+func NewPulseServer() *PulseServer {
+	sessionFactory := env().Database.SessionFactory
+	return &PulseServer{
+		instanceID:    env().Config.MessageBroker.ClientID,
+		pulseInterval: env().Config.PulseServer.PulseInterval,
+		checkInterval: env().Config.PulseServer.CheckInterval,
+		instanceDao:   dao.NewInstanceDao(&sessionFactory),
+		lockFactory:   db.NewAdvisoryLockFactory(sessionFactory),
+		sourceClient:  env().Clients.CloudEventsSource,
+	}
+}
+
+// Start initializes and runs the pulse server.
+// It periodically updates the heartbeat for the current Maestro instance,
+// checks Maestro instances' liveness and trigger statusresync for dead instances.
+func (s *PulseServer) Start(ctx context.Context) {
+	logger := logger.NewOCMLogger(ctx)
+	instance := &api.Instance{
+		Meta: api.Meta{
+			ID:        s.instanceID,
+			UpdatedAt: time.Now(),
+		},
+	}
+	instance, err := s.instanceDao.UpSert(ctx, instance)
+	if err != nil {
+		logger.Error(fmt.Sprintf("Unable to create maestro instance: %s", err.Error()))
+		return
+	}
+
+	pulseTicker := time.NewTicker(time.Duration(s.pulseInterval) * time.Second)
+	checkTicker := time.NewTicker(time.Duration(s.checkInterval) * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			pulseTicker.Stop()
+			checkTicker.Stop()
+			return
+		case <-pulseTicker.C:
+			logger.Infof("Updating heartbeat for maestro instance: %s", instance.ID)
+			instance.UpdatedAt = time.Now()
+			instance, err = s.instanceDao.Replace(ctx, instance)
+			if err != nil {
+				// ignore the error and continue to tolerate the intermittent db connection issue
+				logger.Error(fmt.Sprintf("Unable to update heartbeat for maestro instance: %s", err.Error()))
+				continue
+			}
+		case <-checkTicker.C:
+			logger.Infof("Checking maestro instances liveness and trigger statusresync for dead instances")
+			// Instances not pulsing within the last three check intervals are considered as dead.
+			instances, err := s.instanceDao.All(ctx)
+			if err != nil {
+				// ignore the error and continue to tolerate the intermittent db connection issue
+				logger.Error(fmt.Sprintf("Unable to get all maestro instances: %s", err.Error()))
+				continue
+			}
+			for _, instance := range instances {
+				// Instances not pulsing within the last three check intervals are considered as dead.
+				if !instance.UpdatedAt.After(time.Now().Add(time.Duration(-3*s.checkInterval) * time.Second)) {
+					// lock the Instance with a fail-fast advisory lock context.
+					// this allows concurrent processing of many instances by one or more maestro instances exclusively.
+					lockOwnerID, acquired, err := s.lockFactory.NewNonBlockingLock(ctx, instance.ID, db.Instances)
```

Comment:
I think we may add this lock at the beginning when we check the instance pulse, and we should use a fixed id for advisory lock, e.g. `maestro-instances-pulse-check`

Reply:
I think if once there is an instance failed, we may need to ensure there is only one instance sends resync request to all consumers at the same time

Reply:
updated this to use fix id for lock, and put it at the beginning.